Ability to authenticate with the schema registry via Cognito OAuth

joshuafleck opened the pull request:
- Ability to authenticate with the schema registry via Cognito OAuth
- Will fetch a bearer token to be passed with requests to the schema registry
- If the bearer token expires, it will fetch a new one
- It will retry up to 3 times to fetch a new bearer token before giving up

joshuafleck committed:
Ability to authenticate with the schema registry via Cognito OAuth

- Ability to authenticate with the schema registry via Cognito OAuth
- Will fetch a bearer token to be passed with requests to the schema registry
- If the bearer token expires, it will fetch a new one
- It will retry up to 3 times to fetch a new bearer token before giving up

fbad commented on README.md:
````diff
@@ -274,3 +274,17 @@ end
 # Messaging objects created with the same registry_url will now use the fake server.
 ```
+
+### Cognito (Oauth) Compatability
````
typo

fbad commented on lib/avro_turf/connection_wrapper_with_token_auth.rb:
```diff
@@ -0,0 +1,100 @@
+    result = super
+    @refresh_token_retries_remaining = REFRESH_TOKEN_TRIES
+    result
+  rescue Excon::Error::Unauthorized
```
Can't we call refresh before the request (it's just a if if refresh is not needed, right?)? I think we can reduce requests this way?

fbad commented on lib/avro_turf/messaging.rb:
```diff
@@ -89,6 +95,9 @@ def initialize(
           client_key_data: client_key_data,
+          oauth_url: oauth_url,
+          oauth_client_id: oauth_client_id,
+          oauth_client_secret: oauth_client_secret,
```
lib/avro_turf/messaging.rb#130 shall we add a config param to enable/disable automatic registration of schemas?

joshuafleck committed:
Ability to disable automatic registering of schemas when encoding

This is desireable in case you wish to prevent callers from registering new schemas (intentionally or unintentionally) during encoding.

mijoharas commented on lib/avro_turf/connection_wrapper_with_token_auth.rb:
```diff
@@ -0,0 +1,101 @@
+    @oauth_url = oauth_url
+    @oauth_client_id = oauth_client_id
+    @oauth_client_secret = oauth_client_secret
+    @semaphore = Mutex.new
```
I'm only halfway through reviewing this, so I need to see how this class is used, but this mutex is created per instance. Is this class used as a singleton? or are we trying to protect cross thread activity of the same instance? will come back to this.

joshuafleck replied:
Yes, we maintain one instance of it when we use it, here: https://github.com/meetcleo/meetcleo/pull/23203/files#diff-46c130aaced44d9d3f3a294431826f310567402d64b4b388074ee7a0b9726828R20

joshuafleck replied:
It's so we can take advantage of the caching, here: https://github.com/meetcleo/avro_turf/blob/master/lib/avro_turf/messaging.rb#L79

mijoharas replied:
perfect, I was actually just reviewing that now.

fbad reviewed (approved):
LGTM
